src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlStreamConfigurationProvider.cs

```diff
@@ -40,61 +39,117 @@ public class XmlStreamConfigurationProvider : StreamConfigurationProvider
                 IgnoreWhitespace = true
             };
 
+            XmlConfigurationElement root = null;
+
             using (XmlReader reader = decryptor.CreateDecryptingXmlReader(stream, readerSettings))
             {
-                var prefixStack = new Stack<string>();
-
-                SkipUntilRootElement(reader);
-
-                // We process the root element individually since it doesn't contribute to prefix
-                ProcessAttributes(reader, prefixStack, data, AddNamePrefix);
-                ProcessAttributes(reader, prefixStack, data, AddAttributePair);
+                // keep track of the tree we followed to get where we are (breadcrumb style)
+                var currentPath = new Stack<XmlConfigurationElement>();
 
                 XmlNodeType preNodeType = reader.NodeType;
+
                 while (reader.Read())
                 {
                     switch (reader.NodeType)
                     {
                         case XmlNodeType.Element:
-                            prefixStack.Push(reader.LocalName);
-                            ProcessAttributes(reader, prefixStack, data, AddNamePrefix);
-                            ProcessAttributes(reader, prefixStack, data, AddAttributePair);
+                            var element = new XmlConfigurationElement(reader.LocalName, GetName(reader));
+
+                            if (currentPath.Count == 0)
+                            {
+                                root = element;
+                            }
+                            else
+                            {
+                                var parent = currentPath.Peek();
+
+                                // If parent already has a dictionary of children, update the collection accordingly
+                                if (parent.ChildrenBySiblingName != null)
+                                {
+                                    // check if this element has appeared before, elements are considered siblings if their SiblingName properties match
+                                    if (parent.ChildrenBySiblingName.TryGetValue(element.SiblingName, out var siblings))
+                                    {
+                                        siblings.Add(element);
+                                    }
+                                    else
+                                    {
+                                        parent.ChildrenBySiblingName.Add(element.SiblingName, new List<XmlConfigurationElement> { element });
+                                    }
```

Comment:
I think this would be simpler if written as: 
```suggestion
                                    // check if this element has appeared before, elements are considered siblings if their SiblingName properties match
                                    if (!parent.ChildrenBySiblingName.TryGetValue(element.SiblingName, out var siblings))
                                    {
                                        siblings = new List<XmlConfigurationElement>();
                                        parent.ChildrenBySiblingName.Add(element.SiblingName, siblings);
                                    }
                                    siblings.Add(element);
```

Reply:
Agreed, I've updated it. Thanks!